Feature/pattern fill

oysteinmoseng opened the pull request:
### Added:
- `pattern-fill.js` module that allows setting `color: { pattern: { ... } }` for colors on series and points.
- Nested `series.keys` in Highcharts core that allows setting nested point properties in data arrays, e.g. `keys: ['x', 'value', 'name', 'dataLabels.style.visibility', 'color.pattern.image']`.

There are unit tests and a demo for each, as well as API docs. We should probably update the docs on our website regarding the pattern fills once it is merged in/released.

### Demos:
- Nested keys: `samples/highcharts/series/data-nested-keys`
- Pattern fills: `samples/maps/demo/pattern-fill-map`

### Pattern fill features:
- Set patterns on color, analogous to linear/radial gradients today.
- Per point patterns can be set, and will inherit properties from patterns set on series. This means you can set e.g. a global color for the series patterns, while defining the other properties per point.
- Pattern types supported are `path` and `image`. `path` patterns support all standard SVG path properties.
- 10 default `path` patterns are added (same as in old pattern fill plugin).
- Pattern `color` (for paths) and `opacity` (for both types) are supported, as in the old plugin.
- Patterns can now be offset by `x`/`y` - particularly useful for images.
- Explicit `width`/`height` can be set for both pattern types.
- If an image pattern is set on a point, and `width`/`height` are not set, `width`/`height` will be automatically set to the point bounding box size. If an `aspectRatio` option is given, this ratio will be honored, and the pattern centered within the bounding box.
- An `animation` option can be set on patterns. Image patterns have a short opacity animation on load by default to help reduce flickering.
- An explicit `id` can be provided for the pattern elements, to promote reuse of configurations, but the module will compute its own ID for the pattern automatically, and reuse DOM elements for identical configurations.

jon-a-nygaard commented on js/parts/Point.js:
```diff
@@ -165,6 +165,32 @@ Highcharts.Point.prototype = {
+	setNestedProperty: function (object, value, key) {
+		var nestedKeys = key.split('.'),
+			nestedRet = object;
+		for (var i = 0; i < nestedKeys.length - 1; ++i) {
```
Also `reduce` can be nice to use in cases like `setNestedProperties`:
```javascript
var setNestedProperty = function(object, value, key) {
  var nestedKeys = key.split('.');
  reduce(nestedKeys, function(result, key, i, arr) {
    var isLastKey = arr.length - 1 === i;
    result[key] = (
      isLastKey ?
      value :
      (isObject(result[key]) ? result[key] : {})
    );
    return result[key];
  }, object);
  return object;
}
````

oysteinmoseng replied:
The reduce solution is nice! 👍 I'm not sure it is more readable, but it is shorter so adding it. Also, using `isObject` is smart with the strict parameter, since it also catches cases where the objects are arrays. I tested for a performance difference here, and there doesn't seem to be any in practice, since this is not a bottleneck performance area.

oysteinmoseng commented on js/parts/Point.js:
```diff
@@ -203,7 +229,15 @@ Highcharts.Point.prototype = {
 				// Skip undefined positions for keys
 				if (!keys || options[i] !== undefined) {
-					ret[pointArrayMap[j]] = options[i];
+					if (pointArrayMap[j].indexOf('.') > 0) {
```
Since this is `String.indexOf` it's not a problem here, so just leaving this as is.

jon-a-nygaard replied:
Sorry, did not notice that it was a String, thought it was an Array. 😅

pawelfus commented on js/modules/pattern-fill.src.js:
```diff
@@ -199,6 +282,23 @@ wrap(H.Series.prototype, 'getColor', function (proceed) {
+/**
+ * Set default pattern dimensions on points that have their own color pattern
+ */
+wrap(H.Series.prototype, 'render', function (proceed) {
```
Shouldn't we use `render` event, to limit number `wrap`'s?

oysteinmoseng replied:
Yes ideally, but issues arise with series types that already wrap render. We need this to run before render, but the `render` event happens in `Series.render`, so anything that wraps before render will mess it up. In order for the event system to work we have to move away from wraps entirely in these cases - which I think is the intention eventually - and have a proper priority queue for events.

pawelfus commented on js/modules/pattern-fill.src.js:
```diff
@@ -0,0 +1,320 @@
+
+
+/**
+ * Add the predefined patterns
```
Do we have this in docs somewhere?

oysteinmoseng replied:
I suggest we add it to the docs about colors once we decide to merge this in: https://www.highcharts.com/docs/chart-design-and-style/colors

pawelfus replied:
Sorry, shortcut. I was thinking about API and docs. Like we have [`chart.colors`](https://api.highcharts.com/highcharts/colors).

oysteinmoseng replied:
Aha, the predefined patterns are not in the API currently. Maybe they should be, and should be part of themes. What do you think @TorsteinHonsi?

TorsteinHonsi replied:
Well, it is the same with `linearGradient` and `radialGradient`. I think what we're missing is a programmatic description of the `Color` type. See for example [Point.color](https://api.highcharts.com/class-reference/Highcharts.Point#.color). I think what we need to do is to add a `typedef` for color, then describe the different types.

oysteinmoseng commented:
I added the demos from the old plugin, with some modifications. All options should now be covered, except `pattern.animation` and `pattern.id`.

Also fixed a bug with show/hide series, that would have affected use of patterns with e.g. tooltips as well. The garbage collecting now does not look at points to determine which patterns are in use, but instead goes through the whole chart DOM with `querySelectorAll`.

oysteinmoseng commented:
Browser issues should be fixed now. The Firefox issue was simple, but the IE/Edge problem required a pretty hefty workaround.

Turns out `load` events on SVG image objects are very buggy in MS browsers. To solve this, we now add a dummy HTML image that handles the loading and holds the `load` event handler. Once the HTML image has loaded, we set the src of the SVG image and fire the event handler. In the meantime, the src of the SVG image is just a transparent pixel (encoded as data URI). This allows us to set working onload handlers for SVG image objects across the browser range.
